include/ginkgo/core/matrix/batch_csr.hpp

```diff
@@ -0,0 +1,341 @@
+     */
+    const Csr* apply(ptr_param<const MultiVector<value_type>> b,
+                     ptr_param<MultiVector<value_type>> x) const;
+
+    /**
+     * @copydoc apply(const MultiVector<value_type>*, const
+     * MultiVector<value_type>*, const MultiVector<value_type>*,
+     * MultiVector<value_type>*)
+     */
+    const Csr* apply(ptr_param<const MultiVector<value_type>> alpha,
+                     ptr_param<const MultiVector<value_type>> b,
+                     ptr_param<const MultiVector<value_type>> beta,
+                     ptr_param<MultiVector<value_type>> x) const;
+
+private:
+    /**
+     * Creates an uninitialized Csr matrix of the specified size.
+     *
+     * @param exec  Executor associated to the matrix
+     * @param size  size of the matrix
+     * @param num_nonzeros_per_item  number of nonzeros in each item of the
+     * batch matrix
+     *
+     * @internal It is necessary to pass in the correct nnz_per_item to ensure
+     * that the arrays are allocated correctly. An incorrect value will result
+     * in a runtime failure when the user tries to use any batch matrix
+     * utilities such as create_view_from_item etc.
+     */
+    Csr(std::shared_ptr<const Executor> exec,
+        const batch_dim<2>& size = batch_dim<2>{},
+        size_type num_nonzeros_per_item = {});
+
+    /**
+     * Creates a Csr matrix from an already allocated (and initialized)
+     * array. The column indices array needs to be the same for all batch items.
+     *
+     * @tparam ValuesArray  type of array of values
+     *
+     * @param exec  Executor associated to the matrix
+     * @param size  size of the matrix
+     * @param values  array of matrix values
+     * @param col_idxs  the col_idxs array of a single batch item of the matrix.
+     * @param row_ptrs  the row_ptrs array of a single batch item of the matrix.
+     *
+     * @note If `values` is not an rvalue, not an array of ValueType, or is on
+     *       the wrong executor, an internal copy will be created, and the
+     *       original array data will not be used in the matrix.
+     */
+    template <typename ValuesArray, typename ColIdxsArray,
+              typename RowPtrsArray>
+    Csr(std::shared_ptr<const Executor> exec, const batch_dim<2>& size,
+        ValuesArray&& values, ColIdxsArray&& col_idxs, RowPtrsArray&& row_ptrs)
+        : EnableBatchLinOp<Csr>(exec, size),
+          values_{exec, std::forward<ValuesArray>(values)},
+          col_idxs_{exec, std::forward<ColIdxsArray>(col_idxs)},
+          row_ptrs_{exec, std::forward<RowPtrsArray>(row_ptrs)}
+    {
+        // Ensure that the value and col_idxs arrays have the correct size
+        auto max_num_elems = this->get_common_size()[0] *
+                             this->get_common_size()[1] *
+                             this->get_num_batch_items();
+        GKO_ASSERT(values_.get_size() <= max_num_elems);
```

Comment:
do you need to check this? it is just unused from the batch csr

Reply:
ah, the get_num_stored_elements is based on values.size(), so it is required to only store the batch value